econml/_ortho_learner.py

```diff
@@ -714,6 +787,64 @@ def score(self, Y, T, X=None, W=None, Z=None, sample_weight=None):
     def model_final(self):
         return self._model_final
 
+    @model_final.setter
+    def model_final(self, model_final):
+        self._model_final = clone(model_final, safe=False)
+
+    @property
+    def model_nuisance(self):
+        return self._model_nuisance
+
+    @model_nuisance.setter
+    def model_nuisance(self, model_nuisance):
+        self._model_nuisance = clone(model_nuisance, safe=False)
+        self._cache_invalid_message = "Setting the nuisance model invalidates the cached nuisances"
+
+    @property
+    def n_splits(self):
+        return self._n_splits
+
+    @n_splits.setter
+    def n_splits(self, n_splits):
+        self._n_splits = n_splits
+        self._cache_invalid_message = "Setting the number of splits invalidates the cached nuisances"
+
+    @property
+    def discrete_treatment(self):
+        return self._discrete_treatment
```

Comment:
Do we really need to expose the first stage related arguments and allow users to set something else? I think the value of `refit` is to save computation time when the nuisance models are the same and the data and results are cached. Even they reset a value here they can't only refit final stage but fit the entire pipeline again, it's the same as reinitialize a new estimator and call `fit` in term of computing time and lines of code. Also it's wired to allow user to change the `discrete_treatment` flag since T is fixed and it should be either continuous our categorical. If we only allow user to change final stage specific argument, it could also simplify tons of properties and setters across multiple classes?    

Same for `_RLearner` and classes under DML file.

Reply:
My idea was to make all things that you can set to some value in the initializer also available to update at any other time.  It's true that many of these are not necessary for the refit functionality alone, and also that a workaround would be to create a new instance.

However, from the user perspective I think sklearn's behavior (where you can set any attribute at any time) would be much easier to understand than a policy where you can only set the attributes that don't affect the nuisances, since it is not always completely obvious which attributes affect the nuisances (e.g. the featurizer affects the Y residuals if we use the `linear_first_stages` flag, but doesn't affect any nuisances if it's False).

Eventually, I think we should enable setting attributes even for estimators that don't have support refit (e.g. the metalearners), but for this PR I'm just starting with the OrthoLearner class hierarchy.